lib/rpc/SingleJsonRpcProvider.ts

```diff
@@ -0,0 +1,235 @@
+  }
+
+  private recordError(method: string) {
+    this.healthScore += this.config.ERROR_PENALTY
+    this.log.debug(
+      `${this.url}: method: ${method} error penalty ${this.config.ERROR_PENALTY}, score => ${this.healthScore}`
+    )
+  }
+
+  private recordHighLatency(method: string) {
+    this.healthScore += this.config.HIGH_LATENCY_PENALTY
+    this.log.debug(
+      `${this.url}: method: ${method}, high latency penalty ${this.config.ERROR_PENALTY}, score => ${this.healthScore}`
+    )
+  }
+
+  private recordProviderRecovery(timeInMs: number) {
+    if (this.healthScore === 0) {
+      return
+    }
+    timeInMs = Math.min(timeInMs, this.config.RECOVER_MAX_WAIT_TIME_TO_ACKNOWLEDGE_IN_MS)
+    this.healthScore += timeInMs * this.config.RECOVER_SCORE_PER_MS
+    if (this.healthScore > 0) {
+      this.healthScore = 0
+    }
+    this.log.debug(
+      `${this.url}: healthy: ${this.healthy}, recovery ${timeInMs} * ${this.config.RECOVER_SCORE_PER_MS} = ${
+        timeInMs * this.config.RECOVER_SCORE_PER_MS
+      }, score => ${this.healthScore}`
+    )
+  }
+
+  private checkLastCallPerformance(method: string, perf: SingleCallPerf) {
+    this.log.debug(`checkLastCallPerformance: method: ${method}`)
+    if (!perf.succeed) {
+      metric.putMetric(`${this.metricPrefix}_${method}_FAILED`, 1, MetricLoggerUnit.Count)
+      this.recordError(method)
+    } else if (perf.latencyInMs > this.config.MAX_LATENCY_ALLOWED_IN_MS) {
+      metric.putMetric(`${this.metricPrefix}_${method}_SUCCESS_HIGH_LATENCY`, 1, MetricLoggerUnit.Count)
+      this.recordHighLatency(method)
+    } else {
+      metric.putMetric(`${this.metricPrefix}_${method}_SUCCESS`, 1, MetricLoggerUnit.Count)
+      this.log.debug(`${this.url} method: ${method} succeeded`)
+      // For a success call, we will increase health score.
+      if (perf.startTimestampInMs - this.lastCallTimestampInMs > 0) {
+        this.recordProviderRecovery(perf.startTimestampInMs - this.lastCallTimestampInMs)
+      }
+    }
+    if (this.healthy && this.healthScore < this.config.HEALTH_SCORE_FALLBACK_THRESHOLD) {
+      this.healthy = false
+      this.log.debug(`${this.url} drops to unhealthy`)
+    } else if (!this.healthy && this.healthScore > this.config.HEALTH_SCORE_RECOVER_THRESHOLD) {
+      this.healthy = true
+      this.log.debug(`${this.url} resumes to healthy`)
+    }
+    // No reward for normal operation.
+  }
+
+  evaluateForRecovery() {
+    this.log.debug(`${this.url}: Evaluate for recovery...`)
+    this.getBlockNumber() // Ignore output in the promise
+  }
+
+  // Wrap another layer only for the sake of ease unit testing.
+  // We will test this API to represent the tests of other similar implemented APIs.
+  private _getBlockNumber(): Promise<number> {
+    return super.getBlockNumber()
+  }
+
+  private wrappedFunctionCall(fnName: string, fn: any, ...args: any[]): Promise<any> {
```

Comment:
nit: I think `wrappedFunctionCall` can be slightly more strongly-typed:

`private async wrappedFunctionCall<ReturnType>(fnName: string, fn: (...args: any[]) => Promise<ReturnType>, ...args: any[]): Promise<ReturnType>`